ocs_ci/utility/utils.py

```diff
@@ -1302,25 +1302,104 @@ def get_az_count():
         return 1
 
 
-def ceph_health_check(namespace=None, tries=20, delay=30):
+def ceph_health_check(
+    namespace=None, tries=20, delay=30, intelligent_wait=False,
+    max_timeout_for_pg=600,
+):
     """
     Args:
         namespace (str): Namespace of OCS
             (default: config.ENV_DATA['cluster_namespace'])
         tries (int): Number of retries
         delay (int): Delay in seconds between retries
+        intelligent_wait (bool): If true, the ceph_health is waiting for
+            HEALTH_OK if we have degraded or undersized PGs. In case the time
+            between decrease of degraded/undersized PGs is less than
+            max_timeout_for_pg than we continue to wait.
+        max_timeout_for_pg (int): Max timout in seconds to wait for decrease in
+            degraded/undersized PGs change for intelligent wait mode.
 
     Returns:
         bool: ceph_health_check_base return value with default retries of 20,
             delay of 30 seconds if default values are not changed via args.
 
     """
-    return retry(
-        (CephHealthException, CommandFailed),
-        tries=tries,
-        delay=delay,
-        backoff=1
-    )(ceph_health_check_base)(namespace)
+    if not intelligent_wait:
+        return retry(
+            (CephHealthException, CommandFailed),
+            tries=tries,
+            delay=delay,
+            backoff=1
+        )(ceph_health_check_base)(namespace)
+
+    degraded_pgs = 0
+    undersized_pgs = 0
+    last_decrease = time.time()
+    while tries:
+        decrease_time = None
+        degraded_pgs_diffs = []
+        degraded_pgs_diff_times = []
+        try:
+            return retry(
+                (CommandFailed),
+                tries=tries,
+                delay=delay,
+                backoff=1
+            )(ceph_health_check_base)(namespace)
+        except CephHealthException as ex:
+            str_ex = str(ex)
```

Comment:
This is a very large amount of code to place in an except block. I recommend moving the majority of the logic to a function that you can call from the except block.